File: crash_cost_app/backend/app/metrics.py
"""Pre-defined `primary_metric` registry.

Each metric is a SQL aggregate column name (`numerator_sql`) on the per-geography
row, optionally divided by `denominator` (`population` for per-capita crash
metrics). Census-only metrics (`vehicles_per_person`, `pct_hh_no_vehicle`) are
joined from prepared GeoJSON in :mod:`main` and are not produced by DuckDB.

Aggregate columns (see :mod:`data`):

    n_crashes, n_fatal, n_injury, n_property_damage
    sum_fatalities_person, sum_injuries_person
    sum_cost_total, sum_cost_medical, … sum_cost_cong_propdamage,
    sum_cost_total_economic, sum_cost_qalys
    sum_occupants, sum_nonmotorists, sum_vehicles
    population, vehicles_per_person, pct_hh_no_vehicle  (joined)
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class MetricDef:
    id: str
    label: str
    numerator_sql: str
    format: str  # "currency" | "number" | "rate" | "percent"
    denominator: str | None = None  # None | "population"
    rate_per: float | None = None  # e.g. 100_000 for "per 100k"
    description: str = ""


METRICS: list[MetricDef] = [
    # Totals (crash-aggregated)
    MetricDef(
        id="comprehensive_cost_total",
        label="Total comprehensive cost (2025 USD)",
        numerator_sql="sum_cost_total",
        format="currency",
        description="Sum of crash-level comprehensive costs.",
    ),
    MetricDef(
        id="crashes_total",
        label="Total crashes",
        numerator_sql="n_crashes",
        format="number",
    ),
    MetricDef(
        id="fatalities_total",
        label="Fatalities (person counts)",
        numerator_sql="sum_fatalities_person",
        format="number",
        description="Sum of occupant and non-motorist police injury status = fatal (code 1).",
    ),
    MetricDef(
        id="injuries_total",
        label="Injuries (person counts)",
        numerator_sql="sum_injuries_person",
        format="number",
        description="Sum of occupant and non-motorist injury status codes 2–4 (excludes code 5 / MAIS0).",
    ),
    MetricDef(
        id="medical_cost_total",
        label="Medical costs (2025 USD)",
        numerator_sql="sum_cost_medical",
        format="currency",
    ),
    MetricDef(
        id="congestion_cost_total",
        label="Congestion costs (2025 USD)",
        numerator_sql="sum_cost_congestion",
        format="currency",
    ),
    MetricDef(
        id="property_damage_cost_total",
        label="Property damage costs (2025 USD)",
        numerator_sql="sum_cost_propdamage",
        format="currency",
    ),
    MetricDef(
        id="total_economic_cost_total",
        label="Total economic costs (2025 USD)",
        numerator_sql="sum_cost_total_economic",
        format="currency",
    ),
    MetricDef(
        id="qalys_cost_total",
        label="Cost of life years lost — QALYs (2025 USD)",
        numerator_sql="sum_cost_qalys",
        format="currency",
    ),
    # Per capita (crash sums / population)
    MetricDef(
        id="comprehensive_cost_per_capita",
        label="Total comprehensive cost per capita (2025 USD)",
        numerator_sql="sum_cost_total",
        format="currency",
        denominator="population",
    ),
    MetricDef(
        id="fatalities_per_capita",
        label="Fatalities per 10,000 population",
        numerator_sql="sum_fatalities_person",
        format="rate",
        denominator="population",
        rate_per=10_000,
    ),
    MetricDef(
        id="injuries_per_capita",
        label="Injuries per 10,000 population",
        numerator_sql="sum_injuries_person",
        format="rate",
        denominator="population",
        rate_per=10_000,
    ),
    MetricDef(
        id="medical_cost_per_capita",
        label="Medical costs per capita (2025 USD)",
        numerator_sql="sum_cost_medical",
        format="currency",
        denominator="population",
    ),
    MetricDef(
        id="congestion_cost_per_capita",
        label="Congestion costs per capita (2025 USD)",
        numerator_sql="sum_cost_congestion",
        format="currency",
        denominator="population",
    ),
    MetricDef(
        id="property_damage_cost_per_capita",
        label="Property damage costs per capita (2025 USD)",
        numerator_sql="sum_cost_propdamage",
        format="currency",
        denominator="population",
    ),
    MetricDef(
        id="total_economic_cost_per_capita",
        label="Total economic costs per capita (2025 USD)",
        numerator_sql="sum_cost_total_economic",
        format="currency",
        denominator="population",
    ),
    MetricDef(
        id="qalys_cost_per_capita",
        label="Cost of life years lost (QALYs) per capita (2025 USD)",
        numerator_sql="sum_cost_qalys",
        format="currency",
        denominator="population",
    ),
    # Census context (NHGIS ACS, not crash-derived)
    MetricDef(
        id="vehicles_per_person",
        label="Vehicles per person (census estimate)",
        numerator_sql="vehicles_per_person",
        format="number",
        description="ACS B08201 vehicle stock estimate divided by total population.",
    ),
    MetricDef(
        id="pct_households_no_vehicle",
        label="Percent of households with no vehicle (census)",
        numerator_sql="pct_hh_no_vehicle",
        format="percent",
        description="ACS B08201: percent of households reporting zero vehicles available.",
    ),
]


METRICS_BY_ID = {m.id: m for m in METRICS}

DEFAULT_METRIC_ID = "comprehensive_cost_per_capita"


def compute_metric_value(metric: MetricDef, row: dict) -> float | None:
    """Apply a metric definition to an aggregated row (Python-side)."""
    num = row.get(metric.numerator_sql)
    if num is None:
        return None
    if isinstance(num, float) and num != num:
        return None
    if metric.denominator is None:
        return float(num)
    denom = row.get(metric.denominator)
    if denom in (None, 0) or (isinstance(denom, float) and denom != denom):
        return None
    value = float(num) / float(denom)
    if metric.rate_per is not None:
        value *= metric.rate_per
    return value

File: crash_cost_app/frontend/src/metrics.ts
import type { MetricDef, SummaryFeature } from "./types";

const NUMERATOR_BY_METRIC_ID: Record<string, string> = {
  comprehensive_cost_total: "sum_cost_total",
  crashes_total: "n_crashes",
  fatalities_total: "sum_fatalities_person",
  injuries_total: "sum_injuries_person",
  medical_cost_total: "sum_cost_medical",
  congestion_cost_total: "sum_cost_congestion",
  property_damage_cost_total: "sum_cost_propdamage",
  total_economic_cost_total: "sum_cost_total_economic",
  qalys_cost_total: "sum_cost_qalys",
  comprehensive_cost_per_capita: "sum_cost_total",
  fatalities_per_capita: "sum_fatalities_person",
  injuries_per_capita: "sum_injuries_person",
  medical_cost_per_capita: "sum_cost_medical",
  congestion_cost_per_capita: "sum_cost_congestion",
  property_damage_cost_per_capita: "sum_cost_propdamage",
  total_economic_cost_per_capita: "sum_cost_total_economic",
  qalys_cost_per_capita: "sum_cost_qalys",
  vehicles_per_person: "vehicles_per_person",
  pct_households_no_vehicle: "pct_hh_no_vehicle",
};

/** Mirror of backend compute_metric_value for dashboard charts. */
export function computeMetricValue(
  metric: MetricDef,
  row: SummaryFeature | Record<string, unknown>,
): number | null {
  const numeratorKey = NUMERATOR_BY_METRIC_ID[metric.id];
  if (!numeratorKey) return null;

  const raw = row[numeratorKey];
  if (raw === null || raw === undefined) return null;
  if (typeof raw === "number" && Number.isNaN(raw)) return null;

  if (!metric.denominator) {
    return typeof raw === "number" ? raw : Number(raw);
  }

  const denom = row[metric.denominator];
  if (denom === null || denom === undefined || denom === 0) return null;
  if (typeof denom === "number" && Number.isNaN(denom)) return null;

  let value = Number(raw) / Number(denom);
  if (metric.rate_per != null) value *= metric.rate_per;
  return value;
}

/** Rate metrics shown in the comparison dashboard (per-capita costs + census context). */
const DASHBOARD_CHART_IDS = new Set([
  "comprehensive_cost_per_capita",
  "fatalities_per_capita",
  "injuries_per_capita",
  "medical_cost_per_capita",
  "congestion_cost_per_capita",
  "property_damage_cost_per_capita",
  "total_economic_cost_per_capita",
  "qalys_cost_per_capita",
  "vehicles_per_person",
]);

export function isRateMetric(metric: MetricDef): boolean {
  return metric.denominator === "population";
}

export function rateMetrics(metrics: MetricDef[]): MetricDef[] {
  return metrics.filter((m) => DASHBOARD_CHART_IDS.has(m.id));
}
